### PR TITLE
Prevent dbDelta from running unnecessary ALTER table during upgrade

### PR DESCRIPTION
**Pull request by peterfabian**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

During each request (front end and backend), WC checks whether the version of code that is executed corresponds to the`woocommerce_version` that is stored in the database (function [WC_Install::check_version()](https://github.com/woocommerce/woocommerce/blob/7.5.0/plugins/woocommerce/includes/class-wc-install.php#L257)). If the code's version is newer (higher version), install procedure is executed ([WC_Install::install()](https://github.com/woocommerce/woocommerce/blob/7.5.0/plugins/woocommerce/includes/class-wc-install.php#L377)).

This executes a bunch of functions, and among them is WordPress default `dbDelta` to make sure the db structure is correct and no fatal errors are happening because of a mismatch between what the code expects and what's in the db. 

The problem is that `dbDelta` can't handle all the column types very well. E.g., it fails to see that `BIGINT UNSIGNED` is the same as `bigint(20) unsigned`, but also that `BOOLEAN` is actually `tinyint(1)` in MySQL/MariaDB: 

> "Changed type of wp_woocommerce_sessions.session_id from bigint(20) unsigned to BIGINT UNSIGNED"
> ...

Because of that, `dbDelta` tries to run about 65 `ALTER` queries on each WC update. This creates a problem for infrastructure where WC is managed and/or symlinked, because all WC instances are updated at the same time and they start executing all those meaningless `ALTER` statements that are useless at the same time, bombarding the SQL servers needlessly.

The proposed fix here, until we fix `dbDelta`, is to simply rename the column types so they correspond to what is reported from `DESCRIBE` SQL query and thus `dbDelta` runs nothing, unless it's needed.

Translations carried out:
- `BIGINT UNSIGNED` to `bigint(20) unsigned`
- `TEXT` to `text`
- `VARCHAR(100)` to `varchar(100)`
- `TINYINT(1)` to `tinyint(1)`
- `boolean` to `tinyint(1)`
- `INT` to `int(11)`

I'd also like to note that `DEFAULT '0000-00-00 00:00:00'` is no longer valid `datetime` in newer MySQL versions, but that's tangential to the issue at hand.

<details><summary>Queries for WC 7.5.0</summary>

0: "ALTER TABLE wp_woocommerce_sessions CHANGE COLUMN `session_id` session_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
1: "ALTER TABLE wp_woocommerce_sessions CHANGE COLUMN `session_expiry` session_expiry BIGINT UNSIGNED NOT NULL"
2: "ALTER TABLE wp_woocommerce_api_keys CHANGE COLUMN `key_id` key_id BIGINT UNSIGNED NOT NULL auto_increment"
3: "ALTER TABLE wp_woocommerce_api_keys CHANGE COLUMN `user_id` user_id BIGINT UNSIGNED NOT NULL"
4: "ALTER TABLE wp_woocommerce_attribute_taxonomies CHANGE COLUMN `attribute_id` attribute_id BIGINT UNSIGNED NOT NULL auto_increment"
5: "ALTER TABLE wp_woocommerce_downloadable_product_permissions CHANGE COLUMN `permission_id` permission_id BIGINT UNSIGNED NOT NULL auto_increment"
6: "ALTER TABLE wp_woocommerce_downloadable_product_permissions CHANGE COLUMN `product_id` product_id BIGINT UNSIGNED NOT NULL"
7: "ALTER TABLE wp_woocommerce_downloadable_product_permissions CHANGE COLUMN `order_id` order_id BIGINT UNSIGNED NOT NULL DEFAULT 0"
8: "ALTER TABLE wp_woocommerce_downloadable_product_permissions CHANGE COLUMN `user_id` user_id BIGINT UNSIGNED NULL"
9: "ALTER TABLE wp_woocommerce_downloadable_product_permissions CHANGE COLUMN `download_count` download_count BIGINT UNSIGNED NOT NULL DEFAULT 0"
10: "ALTER TABLE wp_woocommerce_order_items CHANGE COLUMN `order_item_id` order_item_id BIGINT UNSIGNED NOT NULL auto_increment"
11: "ALTER TABLE wp_woocommerce_order_items CHANGE COLUMN `order_item_name` order_item_name TEXT NOT NULL"
12: "ALTER TABLE wp_woocommerce_order_items CHANGE COLUMN `order_id` order_id BIGINT UNSIGNED NOT NULL"
13: "ALTER TABLE wp_woocommerce_order_itemmeta CHANGE COLUMN `meta_id` meta_id BIGINT UNSIGNED NOT NULL auto_increment"
14: "ALTER TABLE wp_woocommerce_order_itemmeta CHANGE COLUMN `order_item_id` order_item_id BIGINT UNSIGNED NOT NULL"
15: "ALTER TABLE wp_woocommerce_tax_rates CHANGE COLUMN `tax_rate_id` tax_rate_id BIGINT UNSIGNED NOT NULL auto_increment"
16: "ALTER TABLE wp_woocommerce_tax_rates CHANGE COLUMN `tax_rate_priority` tax_rate_priority BIGINT UNSIGNED NOT NULL"
17: "ALTER TABLE wp_woocommerce_tax_rates CHANGE COLUMN `tax_rate_order` tax_rate_order BIGINT UNSIGNED NOT NULL"
18: "ALTER TABLE wp_woocommerce_tax_rate_locations CHANGE COLUMN `location_id` location_id BIGINT UNSIGNED NOT NULL auto_increment"
19: "ALTER TABLE wp_woocommerce_tax_rate_locations CHANGE COLUMN `tax_rate_id` tax_rate_id BIGINT UNSIGNED NOT NULL"
20: "ALTER TABLE wp_woocommerce_shipping_zones CHANGE COLUMN `zone_id` zone_id BIGINT UNSIGNED NOT NULL auto_increment"
21: "ALTER TABLE wp_woocommerce_shipping_zones CHANGE COLUMN `zone_order` zone_order BIGINT UNSIGNED NOT NULL"
22: "ALTER TABLE wp_woocommerce_shipping_zone_locations CHANGE COLUMN `location_id` location_id BIGINT UNSIGNED NOT NULL auto_increment"
23: "ALTER TABLE wp_woocommerce_shipping_zone_locations CHANGE COLUMN `zone_id` zone_id BIGINT UNSIGNED NOT NULL"
24: "ALTER TABLE wp_woocommerce_shipping_zone_methods CHANGE COLUMN `zone_id` zone_id BIGINT UNSIGNED NOT NULL"
25: "ALTER TABLE wp_woocommerce_shipping_zone_methods CHANGE COLUMN `instance_id` instance_id BIGINT UNSIGNED NOT NULL auto_increment"
26: "ALTER TABLE wp_woocommerce_shipping_zone_methods CHANGE COLUMN `method_order` method_order BIGINT UNSIGNED NOT NULL"
27: "ALTER TABLE wp_woocommerce_payment_tokens CHANGE COLUMN `token_id` token_id BIGINT UNSIGNED NOT NULL auto_increment"
28: "ALTER TABLE wp_woocommerce_payment_tokens CHANGE COLUMN `user_id` user_id BIGINT UNSIGNED NOT NULL DEFAULT '0'"
29: "ALTER TABLE wp_woocommerce_payment_tokenmeta CHANGE COLUMN `meta_id` meta_id BIGINT UNSIGNED NOT NULL auto_increment"
30: "ALTER TABLE wp_woocommerce_payment_tokenmeta CHANGE COLUMN `payment_token_id` payment_token_id BIGINT UNSIGNED NOT NULL"
31: "ALTER TABLE wp_woocommerce_log CHANGE COLUMN `log_id` log_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
32: "ALTER TABLE wp_wc_webhooks CHANGE COLUMN `webhook_id` webhook_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
33: "ALTER TABLE wp_wc_webhooks CHANGE COLUMN `user_id` user_id BIGINT UNSIGNED NOT NULL"
34: "ALTER TABLE wp_wc_download_log CHANGE COLUMN `download_log_id` download_log_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
35: "ALTER TABLE wp_wc_download_log CHANGE COLUMN `permission_id` permission_id BIGINT UNSIGNED NOT NULL"
36: "ALTER TABLE wp_wc_download_log CHANGE COLUMN `user_id` user_id BIGINT UNSIGNED NULL"
37: "ALTER TABLE wp_wc_download_log CHANGE COLUMN `user_ip_address` user_ip_address VARCHAR(100) NULL DEFAULT ''"
38: "ALTER TABLE wp_wc_tax_rate_classes CHANGE COLUMN `tax_rate_class_id` tax_rate_class_id BIGINT UNSIGNED NOT NULL auto_increment"
39: "ALTER TABLE wp_wc_rate_limits CHANGE COLUMN `rate_limit_id` rate_limit_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
40: "ALTER TABLE wp_wc_rate_limits CHANGE COLUMN `rate_limit_expiry` rate_limit_expiry BIGINT UNSIGNED NOT NULL"
41: "ALTER TABLE wp_wc_product_download_directories CHANGE COLUMN `url_id` url_id BIGINT UNSIGNED NOT NULL auto_increment"
42: "ALTER TABLE wp_wc_product_download_directories CHANGE COLUMN `enabled` enabled TINYINT(1) NOT NULL DEFAULT 0"
43: "ALTER TABLE wp_wc_order_stats CHANGE COLUMN `returning_customer` returning_customer boolean DEFAULT NULL"
44: "ALTER TABLE wp_wc_order_stats CHANGE COLUMN `customer_id` customer_id BIGINT UNSIGNED NOT NULL"
45: "ALTER TABLE wp_wc_order_product_lookup CHANGE COLUMN `order_item_id` order_item_id BIGINT UNSIGNED NOT NULL"
46: "ALTER TABLE wp_wc_order_product_lookup CHANGE COLUMN `order_id` order_id BIGINT UNSIGNED NOT NULL"
47: "ALTER TABLE wp_wc_order_product_lookup CHANGE COLUMN `product_id` product_id BIGINT UNSIGNED NOT NULL"
48: "ALTER TABLE wp_wc_order_product_lookup CHANGE COLUMN `variation_id` variation_id BIGINT UNSIGNED NOT NULL"
49: "ALTER TABLE wp_wc_order_product_lookup CHANGE COLUMN `customer_id` customer_id BIGINT UNSIGNED NULL"
50: "ALTER TABLE wp_wc_order_product_lookup CHANGE COLUMN `product_qty` product_qty INT NOT NULL"
51: "ALTER TABLE wp_wc_order_tax_lookup CHANGE COLUMN `order_id` order_id BIGINT UNSIGNED NOT NULL"
52: "ALTER TABLE wp_wc_order_tax_lookup CHANGE COLUMN `tax_rate_id` tax_rate_id BIGINT UNSIGNED NOT NULL"
53: "ALTER TABLE wp_wc_order_coupon_lookup CHANGE COLUMN `order_id` order_id BIGINT UNSIGNED NOT NULL"
54: "ALTER TABLE wp_wc_order_coupon_lookup CHANGE COLUMN `coupon_id` coupon_id BIGINT NOT NULL"
55: "ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `note_id` note_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
56: "ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_snoozable` is_snoozable boolean DEFAULT 0 NOT NULL"
57: "ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_deleted` is_deleted boolean DEFAULT 0 NOT NULL"
58: "ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_read` is_read boolean DEFAULT 0 NOT NULL"
59: "ALTER TABLE wp_wc_admin_note_actions CHANGE COLUMN `action_id` action_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
60: "ALTER TABLE wp_wc_admin_note_actions CHANGE COLUMN `note_id` note_id BIGINT UNSIGNED NOT NULL"
61: "ALTER TABLE wp_wc_customer_lookup CHANGE COLUMN `customer_id` customer_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT"
62: "ALTER TABLE wp_wc_customer_lookup CHANGE COLUMN `user_id` user_id BIGINT UNSIGNED DEFAULT NULL"
63: "ALTER TABLE wp_wc_category_lookup CHANGE COLUMN `category_tree_id` category_tree_id BIGINT UNSIGNED NOT NULL"
64: "ALTER TABLE wp_wc_category_lookup CHANGE COLUMN `category_id` category_id BIGINT UNSIGNED NOT NULL"

</details>

### How to test the changes in this Pull Request:

- [x] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

#### GUI testing

⚠️ Please note that this method is unreliable and only provided for convenience/experimentation. See below for proper testing. 

This change is a bit tricky to test, because the version check runs on **each request**. Thus, you might need to be a bit lucky to catch it in the GUI. This might work, though, and ideally would be repeated with MySQL 5.5, 5.6, 5.7, <8.0.17, >= 8.0.17 and MariaDB 10.4.26 (the version running on our infra). 

1. Install WP & WC, activate WC (any settings are fine)
2. Install QueryMonitor
3. Using your favourite SQL administration tool (PHP MyAdmin, Adminer, cli mysql) set the `woocommerce_version` to "7.4.0", e.g. by running `UPDATE wp_options SET option_value="7.4.0" WHERE option_name="woocommerce_version"`. Alternatively, you can use WP CLI and run 
```
wp shell
wp> update_option('woocommerce_version', '7.4.0');
```
(please note that `wp option update woocommerce_version '7.4.0'` didn't work for me for some reason, didn't bother checking why).
5. Switch over to your browser and quickly hit refresh
6. Examine the QueryMonitor output for Database Queries. You should see 60+ `ALTER` db queries, similar to the below (showing 1 for illustration purposes):
Screenshot 2023-03-16 at 22 21 47

This problem is currently most pronounced on db server versions except for MySQL >=8.0.17, where you should see around 65 `ALTER` queries, most of them related to conversion of columns to BIGINT UNSIGNED. On MySQL 8.0.17+, you should see around 6 queries, and none of them are related to integer data types.
Screenshot 2023-03-16 at 23 03 21

7. Apply this branch
8. Update `woocommerce_version` again as in step 3.
9. Switch over to your browser and quickly hit refresh
10. Examine the QueryMonitor output for Database Queries. There should be no `ALTER` queries.

#### Programmatic testing

The problem with the GUI approach based on Query Monitor is that some other request might get triggered in the background and we miss the request where the `dbDelta` gets called. Seeing no `ALTER` statements, we'd think all is great, but it's not. Au contraire, a classic instance where the absence of evidence isn't evidence of absence (of `ALTER` statements), so to actually get better, we can add simple logging code to `dbDelta` and examine the logs before the patch after the patch:

1. Install WP & WC, activate WC (any settings are fine)
2. Find the following line in your WordPress installation in file `upgrade.php`: `$allqueries = array_merge( $cqueries, $iqueries );`. In v 6.1.1. it's around line number 3115.
3. Add the following logging code after the line:
```
	$f = fopen('/tmp/queries.txt', 'a');
	$log_record = date("Y-m-d H:i:s") . " " . $_SERVER['REQUEST_URI'] . ": ";
	fwrite($f, $log_record . print_r($allqueries, true));
	fclose($f);
```
4.  Using your favourite SQL administration tool (PHP MyAdmin, Adminer, cli mysql) set the `woocommerce_version` to "7.4.0", e.g. by running `UPDATE wp_options SET option_value="7.4.0" WHERE option_name="woocommerce_version"`. 
5. Go back to the browser and reload the WC page (any page is fine).
6. Check the log file, you should see a bunch of queries from `dbDelta`, e.g. like this on MySQL 8.0.17+
```
2023-03-16 22:15:22 /wp-admin/edit.php?post_type=shop_order: Array
(
    [0] => ALTER TABLE wp_woocommerce_order_items CHANGE COLUMN `order_item_name` order_item_name TEXT NOT NULL
    [1] => ALTER TABLE wp_wc_download_log CHANGE COLUMN `user_ip_address` user_ip_address VARCHAR(100) NULL DEFAULT ''
    [2] => ALTER TABLE wp_wc_order_stats CHANGE COLUMN `returning_customer` returning_customer boolean DEFAULT NULL
    [3] => ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_snoozable` is_snoozable boolean DEFAULT 0 NOT NULL
    [4] => ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_deleted` is_deleted boolean DEFAULT 0 NOT NULL
    [5] => ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_read` is_read boolean DEFAULT 0 NOT NULL
)
```
8. Apply this branch
9. Using your favourite SQL administration tool (PHP MyAdmin, Adminer, cli mysql) set the `woocommerce_version` to "7.4.0", e.g. by running `UPDATE wp_options SET option_value="7.4.0" WHERE option_name="woocommerce_version"`. 
10. Go back to the browser and reload the WC page (any page is fine).
11. Check the log file, you shouldn't see any new queries from `dbDelta`, e.g. like this on MySQL 8.0.17+, first the ALTER queries from previous update, then no queries:
```
2023-03-16 22:15:22 /wp-admin/edit.php?post_type=shop_order: Array
(
    [0] => ALTER TABLE wp_woocommerce_order_items CHANGE COLUMN `order_item_name` order_item_name TEXT NOT NULL
    [1] => ALTER TABLE wp_wc_download_log CHANGE COLUMN `user_ip_address` user_ip_address VARCHAR(100) NULL DEFAULT ''
    [2] => ALTER TABLE wp_wc_order_stats CHANGE COLUMN `returning_customer` returning_customer boolean DEFAULT NULL
    [3] => ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_snoozable` is_snoozable boolean DEFAULT 0 NOT NULL
    [4] => ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_deleted` is_deleted boolean DEFAULT 0 NOT NULL
    [5] => ALTER TABLE wp_wc_admin_notes CHANGE COLUMN `is_read` is_read boolean DEFAULT 0 NOT NULL
)
2023-03-16 22:16:05 /wp-admin/admin-ajax.php: Array
(
)
```

Ideally, we'd test this on our infrastructure (Atomic platform), but also other MySQL servers. Based on local testing and checking results [on db-fiddle](https://www.db-fiddle.com/f/3PMNx6vwxNerug3DDW6Fiy/1), it seems like MySQL versions should be handled correctly, and the MariaDBs I saw as well. 

#### Positive test case:

Edit later:
We should also test a positive test case, e.g. that adding a new column actually still works. 
1. Edit the WC_Install::get_schema and add one column, e.g. add `session_test_column bigint(20) unsigned NOT NULL,` to the very first table so it becomes:
```
CREATE TABLE {$wpdb->prefix}woocommerce_sessions (
  session_id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  session_key char(32) NOT NULL,
  session_value longtext NOT NULL,
  session_expiry bigint(20) unsigned NOT NULL,
  session_test_column bigint(20) unsigned NOT NULL,
  PRIMARY KEY  (session_id),
  UNIQUE KEY session_key (session_key)
) $collate;
```
2. Using your favourite SQL administration tool (PHP MyAdmin, Adminer, cli mysql) set the `woocommerce_version` to "7.4.0", e.g. by running `UPDATE wp_options SET option_value="7.4.0" WHERE option_name="woocommerce_version"`. 
3. Go back to the browser and reload the WC page (any page is fine).
4. Check the log file, you should see a query to create this column:
```
2023-03-17 08:10:04 /wp-admin/admin-ajax.php: Array
(
    [0] => ALTER TABLE wp_woocommerce_sessions ADD COLUMN session_test_column bigint(20) unsigned NOT NULL
)
```

More context: p6q8Tx-2Gk-p2

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by barryhughes** [comment]
The GUI approach with Query Monitor is quite nice. Possible way to make it more reliable would involve a snippet like this, which could be saved to `wp-content/mu-plugins/test-37277.php` or similar:

```php
<?php
/**
 * Plugin name: Test PR#37277
 * Description: https://github.com/woocommerce/woocommerce/pull/37277
 */

// If ?wcver=reset is passed, reset the `woocommerce_version` option to `7.4.0`.
add_action( 'plugins_loaded', function () {
	if ( 'reset' === filter_input( INPUT_GET, 'wcver' ) ) {
		update_option( 'woocommerce_version', '7.4.0' );
	}
}, 0 );
```

Then, add `?wcver=reset` to your URL and reload. No potential for intervening requests to get in the way.

File: plugins/woocommerce/includes/class-wc-install.php [hunk 1]
```diff
@@ -1047,6 +1047,8 @@ public static function create_tables() {
 
 		// Clear table caches.
 		delete_transient( 'wc_attribute_taxonomies' );
+
+		return $db_delta_result;
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-install.php:1053]: Should we cover this with a `@return` tag in the docblock?

File: plugins/woocommerce/tests/legacy/unit-tests/util/install.php [hunk 2]
```diff
@@ -166,4 +166,12 @@ function ( $tables ) {
 
 		$this->assertContains( 'some_table_name', WC_Install::get_tables() );
 	}
+
+	/**
+	 * Test that dbDelta is a noop on an installed site.
+	 */
+	public function test_dbDelta_is_a_noop() {
```
Comment (barryhughes) [inline plugins/woocommerce/tests/legacy/unit-tests/util/install.php]: Thinking of [this guidance](https://github.com/woocommerce/woocommerce/blob/7.5.0/plugins/woocommerce/tests/README.md?plain=1#L130):

> No new tests should be added here, ever; existing test classes shouldn't get new tests either.

Would it be better to move this test to `WC_Install_Test`?

**Review by barryhughes — COMMENTED** [review]
Leaving a couple of minor/initial notes, still need to do more testing.

**Review by vedanshujain — DISMISSED** [review]
Approving, tested on MySQL 5.7 and MariaDB 10.6. Note that I used the following snippet to test out the new schema:

```php
$callback = function() { return self::get_schema(); };
$schema = $callback->call( (new WC_Install()) );
require_once ABSPATH . 'wp-admin/includes/upgrade.php';
$results = dbDelta( $schema );
```

On trunk, this returns a large array of changes like so, even if executed multiple times:
```
> $results = dbDelta( $schema );
= [
    "wp_woocommerce_sessions.session_id" => "Changed type of wp_woocommerce_sessions.session_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_sessions.session_expiry" => "Changed type of wp_woocommerce_sessions.session_expiry from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_api_keys.key_id" => "Changed type of wp_woocommerce_api_keys.key_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_api_keys.user_id" => "Changed type of wp_woocommerce_api_keys.user_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_attribute_taxonomies.attribute_id" => "Changed type of wp_woocommerce_attribute_taxonomies.attribute_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_downloadable_product_permissions.permission_id" => "Changed type of wp_woocommerce_downloadable_product_permissions.permission_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_downloadable_product_permissions.product_id" => "Changed type of wp_woocommerce_downloadable_product_permissions.product_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_downloadable_product_permissions.order_id" => "Changed type of wp_woocommerce_downloadable_product_permissions.order_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_downloadable_product_permissions.user_id" => "Changed type of wp_woocommerce_downloadable_product_permissions.user_id from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_downloadable_product_permissions.download_count" => "Changed type of wp_woocommerce_downloadable_product_permissions.download_count from bigint(20) unsigned to BIGINT UNSIGNED",
    "wp_woocommerce_order_items.order_item_id" => "Changed type of wp_woocommerce_order_items.order_item_id from bigint(20) unsigned to BIGINT UNSIGNED",
.....
```

On this branch, this is empty array:
```
> $results = dbDelta( $schema );
= []
```

I think we can merge after addressing barry's comments, great work!

File: plugins/woocommerce/includes/class-wc-install.php (hunk 1 above)
Comment (peterfabian) [reply plugins/woocommerce/includes/class-wc-install.php:1053]: Good call, added in [bdf1eca](https://github.com/woocommerce/woocommerce/pull/37277/commits/bdf1eca7545c3574cbda7bbc5ccb062d0afe1a1d).

File: plugins/woocommerce/tests/legacy/unit-tests/util/install.php (hunk 2 above)
Comment (peterfabian) [reply plugins/woocommerce/tests/legacy/unit-tests/util/install.php]: True, true. Moved the test in [abe3e28](https://github.com/woocommerce/woocommerce/pull/37277/commits/abe3e28bf1aa73b67a06f9b776aac2b809277ce8).

**Review by vedanshujain — DISMISSED** [review]

**Review by vedanshujain — APPROVED** [review]

**Comment by barryhughes** [comment]
Quick note: the snippet in the programmatic testing steps may not work in environments where writing to `/tmp` is disallowed (or where the system temporary directory is something else). In those cases, we can revise the snippet to:

```php
wc_get_logger()->debug(
	'UPGRADE QUERIES' . PHP_EOL . $_SERVER['REQUEST_URI'] . PHP_EOL . implode( PHP_EOL, $allqueries )
);
```

Then, if you then navigate to **WooCommerce ▸ Status ▸ Logs**, you should see the debug info there:

debug-upgrade-queries

Take note of the timestamp (or else just delete the log) and then re-run the test as needed to see the difference between branches. cc @nigeljamesstevenson

**Pull request merged**